### PR TITLE
[chore] Fixes #42 - Import old wordpress data and convert to markdown…

### LINKED ISSUE #42: Migrate data from old site
I'm assuming we are doing this?

- [x] source dump of MySQL DB
- [x] import posts to news
  - [x] convert html to markdown

### PR DESCRIPTION
**Pull request by benichu**
Left TODOS:
- [x] document execution of rake command on heroku
- [x] cleanup NewsItem model
- [x] controller tests for news_items#show and news_items#index
- [x] unit tests `.published`

PROBLEMS:
- Overriding thoughtbot/administrate does not work as advertised

[Review on Reviewable](https://reviewable.io/reviews/montrealrb/montreal.rb/98)

### CODE REVIEW

**Comment by benichu** [comment]
waiting for #83 to be merged before finishing it up

**Comment by benichu** [comment]
@montrealrb/organizers If anyone of you is interested to do a code-review and merge that PR, it takes care of the following:

- Generate a wordpress database dump in db/legacy.yml
- Create a rake task to import that legacy data in the news_items table
- Add some unit and controller specs for NewsItem
- Deal with legacy urls: /YYYY/mm/a-wordpress-slug

Note: you will noticed that I imported all the posts from wordpress (hopefully) but that I only marked the 10 most recent, because some manual markdown cleanup work will need to be done, and I think we should focus on the most recent posts for launch and after that work our way back...

**Comment by benichu** [comment]
code reviewer, please assign the PR to yourself...

**Comment by benichu** [comment]
Will also Fix #99

File: app/dashboards/news_item_dashboard.rb [hunk 1]
```diff
@@ -26,11 +26,9 @@ class NewsItemDashboard < Administrate::BaseDashboard
   COLLECTION_ATTRIBUTES = [
     :id,
     :title,
-    :body,
     :state,
-    :published_at,
-    :created_at,
     :updated_at,
+    :published_at,
```
Comment (mdeloupy) [inline app/dashboards/news_item_dashboard.rb]: @benichu This array ends with a comma, is that normal ? Coming from you I guess it is, but I've never seen this behavior.

Comment (mdeloupy) [inline app/dashboards/news_item_dashboard.rb]: Got my answer : https://github.com/bbatsov/ruby-style-guide/issues/76

File: lib/tasks/database.rake [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+namespace :database do
+  namespace :legacy do
+    # NOTE to run remotely on heroku:
+    # Source: https://devcenter.heroku.com/articles/rake
+    #
+    # $ heroku run rake database:legacy:import_news
+    #
+    desc "Import news records to database"
+    task :import_news => :environment do
```
Comment (mdeloupy) [inline lib/tasks/database.rake]: Hash rocket

File: lib/tasks/database.rake [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        rescue => e
+          puts news_item.inspect
+        end
+      end
+    end
+
+    # NOTE: This has normally been run by someone with access to the wordpress mysql database dump
+    # The result is `db/legacy.yml`
+    desc "Generate a yml files with legacy news records from wordpress mysql DB"
+    task :dump => :environment do
```
Comment (mdeloupy) [inline lib/tasks/database.rake]: Hash rocket

**Comment by mdeloupy** [comment]
@montrealrb/organizers I did a quick review, but don't want to take the responsibility of merging this... Anyone else up for it ?

File: app/controllers/news_items_controller.rb [hunk 4]
```diff
@@ -1,9 +1,17 @@
...
   def index
-    @news_items = NewsItem.published.to_a
+    @news_items = news_item_scope.to_a
   end
 
   def show
-    @news_item = NewsItem.published.find(params[:id])
+    @news_item = news_item_scope.where(slug: params[:slug]).first ||
+                 news_item_scope.where(id: params[:id]).first ||
+                 raise(ActiveRecord::RecordNotFound)
```
Comment (nicholasjhenry) [inline app/controllers/news_items_controller.rb]: @benichu Could we extract as a method and pushed into the model? e.g.

```ruby
NewsItem.fetch_published!(params[:id], params[:slug])
```

This allow to document the method, as I'm unsure why this is required.

File: db/schema.rb [hunk 5]
```diff
@@ -16,6 +16,25 @@
   # These are extensions that must be enabled in order to support this database
   enable_extension "plpgsql"
 
+  create_table "companies", force: :cascade do |t|
```
Comment (nicholasjhenry) [inline db/schema.rb]: @benichu can you please do a `rake db:migrate:reset` to update your schema? These tables should not be in the schema.

File: app/controllers/news_items_controller.rb (hunk 4 above)
Comment (benichu) [inline app/controllers/news_items_controller.rb]: Sure. The reason why it's required is to still resolved old wordpress links (ex: if someone clicks on an post url from google)
I'll extract that logic into a tested and documented class method tonight or later today...

File: db/schema.rb (hunk 5 above)
Comment (benichu) [inline db/schema.rb]: will do.

File: lib/tasks/database.rake [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    task :import_news => :environment do
+      STDOUT.puts "This will destroy your news_items table. Enter 'Y' to confirm: [y/N]"
+      input = STDIN.gets.chomp
+      raise "Aborting import. You entered #{input}" unless input.downcase == "y"
+
+      NewsItem.destroy_all
+      records = YAML.load_file "#{Rails.root}/db/legacy.yml"
+      records.each do |data|
+        news_item = NewsItem.new
+        published_at = data["post_date"]
```
Comment (nicholasjhenry) [inline lib/tasks/database.rake]: @benichu Since there's a couple of changes that need to be made, how would you feel call #fetch on these?

```
 data.fetch("post_date")
```

I like exceptions thrown when the key I expect doesn't exist.

File: lib/tasks/database.rake [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+        news_item.state        = published_at > DateTime.new(2015, 1, 1).beginning_of_year ? :published : :archived
```
Comment (nicholasjhenry) [inline lib/tasks/database.rake]: I added the state `archived` to the domain model. Perhaps we don't actually need this. Thoughts?

File: lib/tasks/database.rake (hunk 6 above)
Comment (benichu) [inline lib/tasks/database.rake]: `published_at` could be nil, if the original post is a draft (which will not happen with the collection I am parsing, as the original query generated only published posts with a date). So it's not really necessary to catch the non-existence of that key.

NOTE: It's a one-off task anyway

File: lib/tasks/database.rake (hunk 7 above)
Comment (benichu) [inline lib/tasks/database.rake]: The only reason why I am doing this, is because I wanted to automatically mark as published the 2015 posts, leaving us time to clean up old posts one by one.

File: app/controllers/news_items_controller.rb (hunk 4 above)
Comment (nicholasjhenry) [inline app/controllers/news_items_controller.rb]: Gotcha, we probably should do a redirect to the correct URL, but let's do that as a separate PR.

File: lib/tasks/database.rake (hunk 6 above)
Comment (nicholasjhenry) [inline lib/tasks/database.rake]: Not quite what I meant -- the key should exists, but the value might be nil :-) I should have clarified that this would be for all keys we're accessing in the hash. But happy to move on given it's a one-off task.

File: lib/tasks/database.rake (hunk 7 above)
Comment (nicholasjhenry) [inline lib/tasks/database.rake]: Perfect.

**Comment by nicholasjhenry** [comment]
Awesome, @benichu! I ran this locally -- looks like we got us a site :-)

**Comment by benichu** [comment]
Cool, I'll do a couple of the fixes you mentioned and we are good to merge, talk to you later...

File: app/controllers/news_items_controller.rb (hunk 4 above)
Comment (benichu) [inline app/controllers/news_items_controller.rb]: You are right, just created a ticket for that #151
Until then, I am also going to leave the object loading methods as-is in the controller, I'll just add a comment to the method.

**Comment by benichu** [comment]
@nicholasjhenry rebased, I'll merge tomorrow morning.

**Comment by sophiedeziel** [comment]
Haha you have to rebase again :stuck_out_tongue:
